lumen/ai/agents.py

```diff
@@ -826,55 +730,71 @@ async def _finalize_execution(
         results: dict,
         context_entries: list[dict],
         messages: list[Message],
+        context: TContext,
         step_title: str | None,
         raise_if_empty: bool = False
-    ) -> None:
+    ) -> tuple[LumenOutput, SQLOutputs]:
         """Finalize execution for final step."""
+
         # Get first result (typically only one for final step)
         expr_slug, result = next(iter(results.items()))
 
-        # Update memory
+        # Update context
         pipeline = result["pipeline"]
         df = await get_data(pipeline)
 
         # If dataframe is empty, raise error to fall back to exploration
         if df.empty and raise_if_empty:
             raise ValueError(f"\nQuery `{result['sql']}` returned empty results; ensure all the WHERE filter values exist in the dataset.")
 
-        self._memory["data"] = await describe_data(df)
-        self._memory["sql"] = result["sql"]
-        self._memory["pipeline"] = pipeline
-        self._memory["table"] = pipeline.table
-        self._memory["source"] = pipeline.source
-        self._memory["sql_plan_context"] = context_entries
-
-        # Render output
-        self._render_lumen(
-            pipeline,
-            messages=messages,
-            title=step_title,
-            spec=result["sql"]
+        view = self._output_type(
+            component=pipeline, title=step_title, spec=result["sql"]
         )
+        return view, {
+            "data": await describe_data(df),
+            "sql": result["sql"],
+            "pipeline": pipeline,
+            "table": pipeline.table,
+            "source": pipeline.source,
+            "sql_plan_context": context_entries
+        }
+
+    def _merge_sources(
+        self, sources: dict[tuple[str, str], BaseSQLSource], tables: list[tuple[str, str]]
+    ) -> tuple[BaseSQLSource, list[str]]:
+        if len(set(m.source for m in tables)) == 1:
+            m = tables[0]
+            return sources[(m.source, m.table)], [m.table]
+        mirrors = {}
+        for m in tables:
+            if not any(m.table.rstrip(")").rstrip("'").rstrip('"').endswith(ext)
+                       for ext in [".csv", ".parquet", ".parq", ".json", ".xlsx"]):
+                renamed_table = m.table.replace(".", "_")
```

Comment:
If I vaguely remember correctly, we have these extensions littered across the codebase (e.g. coordinator.py or ui.py); we should have a constant for the recognized ones.

I feel like we should have a better way to handle this, i.e. util or have source include a method to do this.